### ISSUE TITLE
[RFC]: Ming-flash-omni-2.0 Talker Perf Optimization Roadmap

### ISSUE THREAD

**Issue by yuanheng-zhao** [vllm-project/vllm-omni#4129]
### Motivation.

Fix concurrency issues and add perf optimization features, to improve throughput of Ming-flash-omni-2.0 talker stage (Ming TTS).

### Proposed Change.

Ming-flash-omni2.0 Talker Stage Perf Tracks:

*The following table will be continuously updated. Please left your ideas and let's discuss under this RFC. I'll suggest to start from triaging, testing, and verifying a specific subtask, and then create a PR to resolve one or multiple related subtask(s). 

| Track | Conditions relate | Key Code/Edit Scope (for reference-only) | Status | Assignee |
|---|---|---|---|---|
| CUDA graph | enable/revise CFMGraphExecutor | check the impl & usage of `CFMGraphExecutor`, yaml `enforce_eager` | WIP #4060 | @Sy0307  |
| AR decode-loop sync & RNG | per-step D2H sync; wasted RNG/timestep | `talker_module.py` | WIP #4060 | @Sy0307  |
| Cross-request batching | max batch size 1 -> concurrency queues | StaticCache, CFM pool, loop | WIP #4060 | @Sy0307  |
| VAE streaming decode | redundant re-decode | check and revise stream_decode in `audio_vae.py`, maybe try `torch.compile` decode | Partially in #4060 | @Sy0307  |
| ~~Attention backend~~ | ~~check and compare SDPA vs FA, revise FA consistently~~ | ~~Qwen2 LLM `_attn_implementation`~~ | ~~**Todo**~~ |   |
| Sampler tuning | CFG variants | `config.steps`/`cfg_strength`, `get_epss_timesteps`, `forward_with_cfg`, etc | Todo: tune & test |  |
| Prompt/prefix reuse | per-segment re-prefill + re-tokenize | `build_tts_input`, `_generate_latents`; maybe refer to VoxCPM2/Qwen3-TTS | Todo | @leohuang257  |
| Incremental audio streaming | `forward()` returns whole waveform, TTFP ≈ E2E for now | `forward`/`_decode_to_output` + serving SSE (`serving_speech.py`) | **Todo** | @LHXuuu  |
| LLM decode-step graph/compile | LLM decode step runs eager for now | `llm_step` (talker_module.py) | Todo | @LHXuuu |
| Port talker to vLLM-native paged LLM | currently use HF `Qwen2Model` + `StaticCache` | talker model backbone; expect to adapt attention backend as well | Todo | @BruceLoveDecimal  |

Bridging of multiple stages
*The following are for mainly for Ming-flash-omni-2.0 thinker + X stages
| Track | Conditions relate | Key Code/Edit Scope (for reference-only) | Status | Assignee |
|---|---|---|---|---|
| Async chunk (thinker -> talker)	| for now full-payload hand-off; we shall enable streaming thinker->talker input to cut omni-speech TTFP | deploy yaml async_chunk: true, thinker2talker path | WIP #4037 | @yuanheng-zhao  |
| Async chunk (thinker -> image-gen)	| Same as above but out of scope of Ming-TTS | thinker2imagegen path	| Todo |  |

### CC List.

cc @ZhengWG , @LHXuuu , @Sy0307 , @R2-Y , @david6666666 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by leohuang257** [comment]
hi @yuanheng-zhao, are these open for contributions? can i take **Attention backend** and **Prompt/prefix reuse**?

**Comment by LHXuuu** [comment]
@yuanheng-zhao I can take Incremental audio streaming and LLM decode-step graph/compile

**Comment by yuanheng-zhao** [comment]
> hi [@yuanheng-zhao](https://github.com/yuanheng-zhao), are these open for contributions? can i take **Attention backend** and **Prompt/prefix reuse**?

Hey @leohuang257 , thanks for contributing! I've added you as an assignee of the subtasks. 

And I suggest to start from checking existing features/impl in mostly optimized VoxCPM2 and Qwen3-TTS (if available), verify, and triage specific patterns on Ming.

**Comment by yuanheng-zhao** [comment]
> [@yuanheng-zhao](https://github.com/yuanheng-zhao) I can take Incremental audio streaming and LLM decode-step graph/compile

Hey @LHXuuu, table updated with your subtasks. We could start from checking existing features/impl in mostly optimized VoxCPM2 and Qwen3-TTS if available.

**Comment by Sy0307** [comment]
Thanks for summary from @yuanheng-zhao and welcome anybody to collaborate with me to for better total performance.

**Comment by BruceLoveDecimal** [comment]
Hi @yuanheng-zhao , I have looked into the Talker part, and my understanding of “Port talker to vLLM-native paged LLM” is following:
```bash
Ming Talker
  ├─ Qwen2 LLM backbone attention   <-- main target of native paged LLM port
  ├─ CFM / DiT attention            <-- not part of the paged LLM scope for now
  └─ AudioVAE decoder attention     <-- also not the main scope of this issue
```
```bash
Current Talker decode loop
  -> HF Qwen2Model
       -> HF attention
       -> HF StaticCache
  -> CFM / DiT
  -> Aggregator
  -> VAE decoder
```
Target
```bash
vLLM AR scheduler
  -> Talker preprocess provides inputs_embeds
  -> vLLM Qwen2Model
       -> vLLM Attention / PagedAttention
       -> vLLM paged KV cache
  -> Talker generates audio latent
  -> Aggregator generates inputs_embeds for the next step
  -> fake logits tell the scheduler whether to continue or stop
```
I’m not sure whether my understanding is fully accurate, so please let me know if I’m missing anything.
If this task is not urgent, could you assign it to me? I may need around 1-2 weeks to complete it.

**Comment by leohuang257** [comment]
@yuanheng-zhao After studying VoxCPM2 and Qwen3-TTS, I'd like to confirm the scope of my two tracks before investing, since both seem largely tied to the vLLM-native port:

**1. Attention backend**
The only lever here is `_attn_implementation='sdpa'`, a flag on the HF `Qwen2Model` (`ming_flash_omni_talker.py:108`, backbone built at `:116`). If the talker is ported to vLLM-native, that whole HF object goes away. Both ported references carry no such flag — Qwen3-TTS uses vLLM `Qwen3Model` (`qwen3_tts_talker.py:311`) and VoxCPM2 uses vLLM `Attention`/PagedAttention (`voxcpm2/minicpm4_paged.py:82`) — the attention backend is engine-selected at runtime. So this track looks redundant / subsumed by the port. Should it be closed or folded into the port? 

**2. Prompt/prefix reuse**
The per-segment loop (`ming_flash_omni_talker.py:453-478`) opens a fresh `StaticCache` per segment (`_init_kv_cache`, `talker_module.py:1070`) and re-prefills the shared prefix every time. This re-prefill half looks like it should be handled by vLLM prefix caching the way VoxCPM2 and Qwen3-TTS do (i.e. setting `enable_prefix_caching: true` in the yaml), rather than a hand-rolled StaticCache cache. The only part outside the port's scope here is the host-side re-tokenize memoization in `build_tts_input` (`talker_module.py:810-823`).

Could you confirm this understanding before I go deeper? I'll follow up on where to best contribute afterwards.

**Comment by yuanheng-zhao** [comment]
> Hi [@yuanheng-zhao](https://github.com/yuanheng-zhao) , I have looked into the Talker part, and my understanding of “Port talker to vLLM-native paged LLM” is following:
> 
> Ming Talker
>   ├─ Qwen2 LLM backbone attention   <-- main target of native paged LLM port
>   ├─ CFM / DiT attention            <-- not part of the paged LLM scope for now
>   └─ AudioVAE decoder attention     <-- also not the main scope of this issue
> ...
> I’m not sure whether my understanding is fully accurate, so please let me know if I’m missing anything. If this task is not urgent, could you assign it to me? I may need around 1-2 weeks to complete it.

Hey @BruceLoveDecimal, thanks for contributing! Your understanding and direction is mostly correct, plus a few additional notes:

1. Consider the decode loop as Qwen2Model -> CFM/DiT -> Aggregator -> stop_head with AudioVAE runs once after the loop. 
2. I think the actual efforts will be determining and verifying if we need any adaptor or wrapper to vllm layers for Ming talker usage, and check how to orchestrate request states, handle synthetic stop logits, etc. And some of the logics might be overlapped and done by #4060 already.  

I've assigned this subtask to you and you could refer to the implementations from `VoxCPM2` first.

**Comment by yuanheng-zhao** [comment]
> [@yuanheng-zhao](https://github.com/yuanheng-zhao) After studying VoxCPM2 and Qwen3-TTS, I'd like to confirm the scope of my two tracks before investing, since both seem largely tied to the vLLM-native port:
> 
> **1. Attention backend** ...
> 
> Could you confirm this understanding before I go deeper? I'll follow up on where to best contribute afterwards.

Hey @leohuang257 , thanks for triaging. Here are some notes:

1. Agreed. Ming's talker flow is simlilar to that of VoxCPM2. After vllm native modules/layers ported it's expected to be selected at runtime rather than manually applying vllm-omni custom wrapper of FA. Let me merge the two related subtasks.
2. Yeah it's expected to deprecate `StaticCache`, and it might be a little different to consider per segment prefix cache re-use from llm's.

**Comment by yuanheng-zhao** [comment]
Hey @leohuang257 , any updates on the Prompt/prefix reuse?

**Comment by LHXuuu** [comment]
PR #4341 — Ming-omni-tts MoE (16.8B-A3B) support — has been merged. We can now start performance optimization and accuracy alignment. @yuanheng-zhao @Sy0307 @leohuang257 @BruceLoveDecimal

**Comment by leohuang257** [comment]
Hey @yuanheng-zhao, sorry for the delay on this. Looking back I think I misread your earlier reply, when you said "let me merge the two related subtasks" I took it to mean both my tracks were being folded into the port, so I deprioritized this and let it slip. My mistake. I had a quick look and my current thinking is that Prompt/prefix reuse should land on top of @BruceLoveDecimal's native-paged port #4284, as a cache_salt covering the conditioning plus front-loading the reference latent, referencing how Qwen3-TTS does it. I'll provide a more detailed update ASAP.

**Comment by Sheehan20** [comment]
Hi @yuanheng-zhao, is Sampler tuning still open? I'd like to take it.

One thing to confirm — should this target ming_flash_omni or ming_tts? You
mentioned in #4942 that the scope here would move to Ming-omni-tts.
